cdap-client/src/main/java/co/cask/cdap/client/ProgramClient.java

```
@@ -353,7 +355,18 @@ public void stopAll(NamespaceId namespace)
           ProgramId program = appId.program(programRecord.getType(), programRecord.getName());
           String status = this.getStatus(program);
           if (!status.equals("STOPPED")) {
-            this.stop(program);
+            try {
+              this.stop(program);
+            } catch (IOException ioe) {
+              // ProgramClient#stop calls RestClient, which throws an IOException if the HTTP response code is 400,
+              // which can be due to the program already being stopped when calling stop on it.
+              if (!"STOPPED".equals(getStatus(program))) {
+                throw ioe;
+              }
+              // Most likely, there was a race condition that the program stopped between the time we checked its
+              // status and calling the stop method.
+              LOG.info("Got IOException while stopping program.", ioe);
```

This should be a WARN.

---

We should also mention that we are ignoring this error

```"Program {} is already stopped, proceeding further even though the following exception is raised", programName, ioE```

---

Updated the code to reflect these suggestions.